### PR TITLE
Add relationships in Tabular connectors

### PR DESCRIPTION
**Pull request by LucGenetier**
Add tests for SalesForce and ZenDesk tabular connectors
Add lazy/recursive TabularRecordType

### CODE REVIEW

File: src/libraries/Microsoft.PowerFx.Connectors/ConnectorFunction.cs [hunk 1]
```diff
@@ -1032,11 +1028,78 @@ private static ConnectorType GetConnectorTypeInternal(ConnectorCompatibility com
...
+                                                         })
+                                                         .ToList();
+
+                connectorType.AddReferencedEntities(refEntities);
             }
 
             return connectorType;
         }
+        
+        public class ReferencedEntity
```
Comment (LucGenetier) [inline src/libraries/Microsoft.PowerFx.Connectors/ConnectorFunction.cs:1055]: >ReferencedEntity [](http://example.com/codeflow?start=21&length=16)

need to move in different file

File: src/libraries/Microsoft.PowerFx.Connectors/OpenApiExtensions.cs [hunk 2]
```diff
@@ -701,7 +737,9 @@ internal static ConnectorType GetConnectorReturnType(this OpenApiOperation openA
             if (response.Content.Count == 0)
             {
                 OpenApiSchema schema = new OpenApiSchema() { Type = "string", Format = "no_format" };
-                return new OpenApiParameter() { Name = "response", Required = true, Schema = schema, Extensions = response.Extensions }.GetConnectorType(compatibility);
+                //return new OpenApiParameter() { Name = "response", Required = true, Schema = schema, Extensions = response.Extensions }.GetConnectorType(compatibility);
```
Comment (LucGenetier) [inline src/libraries/Microsoft.PowerFx.Connectors/OpenApiExtensions.cs:740]: >//return new OpenApiParameter() { Name = "response", Required = true, Schema = schema, Extensions = response.Extensions }.GetConnectorType(compatibility); [](http://example.com/codeflow?start=16&length=154)

need to remove

File: src/libraries/Microsoft.PowerFx.Connectors/Internal/Wrappers/ConnectorApiDiscriminator.cs [hunk 3]
```diff
@@ -0,0 +1,24 @@
+﻿// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT license.
+
+using Microsoft.OpenApi.Models;
+
+namespace Microsoft.PowerFx.Connectors
+{
+    internal class ConnectorApiDiscriminator : IConnectorDiscriminator
```
Comment (LucGenetier) [inline src/libraries/Microsoft.PowerFx.Connectors/Internal/Wrappers/ConnectorApiDiscriminator.cs:8]: >ConnectorApiDiscriminator [](http://example.com/codeflow?start=19&length=25)

rename

File: src/libraries/Microsoft.PowerFx.Connectors/Internal/Wrappers/ConnectorApiExtensions.cs [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+
+        public ConnectorApiExtensions(IDictionary<string, IOpenApiExtension> extensions)
+        {
+            _extensions = extensions;
+        }
+
+        public IDictionary<string, IOpenApiExtension> Extensions => _extensions ?? new Dictionary<string, IOpenApiExtension>();
+    }
+
+    internal class ConnectorJsonExtensions : IConnectorExtensions
```
Comment (LucGenetier) [inline src/libraries/Microsoft.PowerFx.Connectors/Internal/Wrappers/ConnectorApiExtensions.cs:36]: >ConnectorJsonExtensions [](http://example.com/codeflow?start=19&length=23)

isolate

File: src/libraries/Microsoft.PowerFx.Connectors/Internal/Wrappers/ConnectorApiDiscriminator.cs (hunk 3 above)
Comment (LucGenetier) [inline src/libraries/Microsoft.PowerFx.Connectors/Internal/Wrappers/ConnectorApiDiscriminator.cs:8]: >ConnectorApiDiscriminator [](http://example.com/codeflow?start=19&length=25)

need parallel testing w/ Json

File: src/libraries/Microsoft.PowerFx.Connectors/Public/ConnectorRelationships.cs [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+namespace Microsoft.PowerFx.Connectors
+{
+    public class ConnectorRelationships
+    {
+        internal ConnectorType _connectorType;
+
+        // Fields that have a relationship with another table
+        public List<ReferencedEntity> FieldsWithRelationship { get; }
+
+        // External tables having a relationship with this table (could be be circular)
```
Comment (LucGenetier) [inline src/libraries/Microsoft.PowerFx.Connectors/Public/ConnectorRelationships.cs:18]: >be [](http://example.com/codeflow?start=75&length=2)

typo

File: src/libraries/Microsoft.PowerFx.Connectors/Tabular/Services/ConnectorServiceBase.cs [hunk 6]
```diff
@@ -26,8 +27,10 @@ protected internal async Task<T> GetObject<T>(HttpClient httpClient, string mess
             return string.IsNullOrWhiteSpace(result) ? null : JsonSerializer.Deserialize<T>(result);
         }
 
+        //static int __i = 0;
```
Comment (LucGenetier) [inline src/libraries/Microsoft.PowerFx.Connectors/Tabular/Services/ConnectorServiceBase.cs:30]: >//static int __i = 0; [](http://example.com/codeflow?start=8&length=21)

remove

File: src/tests/Microsoft.PowerFx.Connectors.Tests.Shared/PowerPlatformTabularTests.cs [hunk 7]
```diff
@@ -442,21 +451,109 @@ public async Task SP_CdpTabular()
...
+            string expr = @"First(Filter(Accounts, 'Account ID' = ""001DR00001Xlq74YAB"")).'Account Name'";
+            CheckResult check = engine.Check(expr, options: new ParserOptions() { AllowsSideEffects = true }, symbolTable: symbolValues.SymbolTable);
+            Assert.True(check.IsSuccess);
+
+            testConnector.SetResponseFromFile(@"Responses\SF GetData.json");
+            FormulaValue result = await check.GetEvaluator().EvalAsync(CancellationToken.None, rc);
+            Assert.Equal("Kutch and Sons", ((StringValue)result).Value);
+        }
+
+        [Fact(Skip = "Under development")]
```
Comment (LucGenetier) [inline src/tests/Microsoft.PowerFx.Connectors.Tests.Shared/PowerPlatformTabularTests.cs:541]: >[Fact(Skip = "Under development")] [](http://example.com/codeflow?start=8&length=34)

remove Skip

File: src/tests/Microsoft.PowerFx.Connectors.Tests.Shared/PowerPlatformTabularTests.cs [hunk 8]
```diff
@@ -491,15 +589,18 @@ public async Task SF_CdpTabular_GetTables()
...
+                "Modified By ID'[User]:s, LastModifiedDate`'Last Modified Date':d, LastReferencedDate`'Last Referenced Date':d, LastViewedDate`'Last Viewed Date':d, MasterRecordId`'Master Record ID'[Account]:s, Name`'Account " +
+                "Name':s, NumberOfEmployees`Employees:w, OwnerId`'Owner ID'[User]:s, ParentId`'Parent Account ID'[Account]:s, Phone`'Account Phone':s, PhotoUrl`'Photo URL':s, ShippingCity`'Shipping City':s, ShippingCountry`'Shipping " +
+                "Country':s, ShippingGeocodeAccuracy`'Shipping Geocode Accuracy':s, ShippingLatitude`'Shipping Latitude':w, ShippingLongitude`'Shipping Longitude':w, ShippingPostalCode`'Shipping Zip/Postal Code':s, ShippingState`'Shipping " +
+                "State/Province':s, ShippingStreet`'Shipping Street':s, SicDesc`'SIC Description':s, SystemModstamp`'System Modstamp':d, Type`'Account Type':s, Website:s]", sfTable.ToStringWithDisplayNames());
 
             HashSet<IExternalTabularDataSource> ads = sfTable.Type._type.AssociatedDataSources;
```
Comment (LucGenetier) [inline src/tests/Microsoft.PowerFx.Connectors.Tests.Shared/PowerPlatformTabularTests.cs:599]: >HashSet [](http://example.com/codeflow?start=12&length=7)

Add relationships testing

File: src/Directory.Build.props [hunk 9]
```diff
@@ -31,10 +31,10 @@
     <EnableNETAnalyzers>True</EnableNETAnalyzers>
     <AnalysisLevel>8.0.0</AnalysisLevel>
     <EnforceCodeStyleInBuild>true</EnforceCodeStyleInBuild>
-    <TreatWarningsAsErrors>true</TreatWarningsAsErrors>
+    <TreatWarningsAsErrors Condition="('$(Configuration)' == 'Release') Or ('$(Configuration)' == 'Release462') Or ('$(Configuration)' == 'Release70') Or ('$(Configuration)' == 'ReleaseAll')">true</TreatWarningsAsErrors>
```
Comment (MikeStall) [inline src/Directory.Build.props:34]: why any changes here?

File: src/tests/Microsoft.PowerFx.Connectors.Tests.Shared/PublicSurfaceTests.cs [hunk 10]
```diff
@@ -50,13 +50,17 @@ public void PublicSurfaceTest_Connectors()
...
               "Microsoft.PowerFx.Connectors.RuntimeConnectorContextExtensions",
               "Microsoft.PowerFx.Connectors.SupportsConnectorErrors",
-              "Microsoft.PowerFx.Connectors.Tabular.ConnectorDataSource",
-              "Microsoft.PowerFx.Connectors.Tabular.ConnectorServiceBase",
-              "Microsoft.PowerFx.Connectors.Tabular.ConnectorTable",
-              "Microsoft.PowerFx.Connectors.Tabular.TabularService",
+              "Microsoft.PowerFx.Connectors.TabularDataSource",
+              "Microsoft.PowerFx.Connectors.TabularRecordType",
+              "Microsoft.PowerFx.Connectors.TabularService",
+              "Microsoft.PowerFx.Connectors.TabularServiceBase",
```
Comment (MikeStall) [inline src/tests/Microsoft.PowerFx.Connectors.Tests.Shared/PublicSurfaceTests.cs:60]: why this distinction? Ie, is is meaningful to expose both base and derived?

File: src/libraries/Microsoft.PowerFx.Connectors/Public/TabularTableValue.cs [hunk 11]
```diff
@@ -14,22 +13,26 @@ namespace Microsoft.PowerFx.Connectors
 {
     // Created by TabularService.GetTableValue
     // Doesn't contain any ServiceProvider which is runtime only
-    public class ConnectorTableValue : TableValue, IRefreshable, IDelegatableTableValue
+    public class TabularTableValue : TableValue, IRefreshable, IDelegatableTableValue
```
Comment (MikeStall) [inline src/libraries/Microsoft.PowerFx.Connectors/Public/TabularTableValue.cs:16]: >TabularTableValue [](http://example.com/codeflow?start=17&length=17)

Can we keep old names?
"ConnectorTable" sounds better than "TabularTable"

File: src/libraries/Microsoft.PowerFx.Connectors/Public/TabularRecordType.cs [hunk 12]
```diff
@@ -0,0 +1,82 @@
...
+            try
+            {
+                TabularTableDescriptor ttd = TableResolver.ResolveTableAsync(tableName, CancellationToken.None).ConfigureAwait(false).GetAwaiter().GetResult();
+
+                type = ttd.ConnectorType.FormulaType;
+                return true;
+            }
+            catch (Exception ex)
+            {
+                TableResolver?.Logger.LogException(ex, $"Cannot resolve external table {tableName}");
```
Comment (MikeStall) [inline src/libraries/Microsoft.PowerFx.Connectors/Public/TabularRecordType.cs:55]: swallowing exception here is dangerous. 
At the least, return an ErrorValue (not blank)

File: src/libraries/Microsoft.PowerFx.Connectors/Public/TabularRecordType.cs [hunk 13]
```diff
@@ -0,0 +1,82 @@
...
+                return _type.LazyTypeProvider.BackingFormulaType.Equals(otherRecordType._type.LazyTypeProvider.BackingFormulaType);
+            }
+
+            return _type.Equals(otherRecordType._type);
+        }
+
+        public override int GetHashCode()
+        {
+            return _type.GetHashCode();
+        }
```
Comment (MikeStall) [inline src/libraries/Microsoft.PowerFx.Connectors/Public/TabularRecordType.cs:80]: do we need this?

Equality on records is ill-defined, so it's a red flag is anybody is every putting these in a hash . 
Almost better to just impl and throw NotSupported

File: src/libraries/Microsoft.PowerFx.Connectors/Public/ReferencedEntity.cs [hunk 14]
```diff
@@ -0,0 +1,14 @@
+﻿// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT license.
+
+namespace Microsoft.PowerFx.Connectors
+{
+    public class ReferencedEntity
+    {
+        public string FieldName { get; init; }
+
+        public string RelationshipName { get; init; }
```
Comment (MikeStall) [inline src/libraries/Microsoft.PowerFx.Connectors/Public/ReferencedEntity.cs:10]: >string [](http://example.com/codeflow?start=15&length=6)

what is this?

File: src/libraries/Microsoft.PowerFx.Connectors/Public/ConnectorRelationships.cs [hunk 15]
```diff
@@ -0,0 +1,36 @@
+﻿// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT license.
+
+using System.Collections.Generic;
+using System.Linq;
+
+namespace Microsoft.PowerFx.Connectors
+{
+    public class ConnectorRelationships
```
Comment (MikeStall) [inline src/libraries/Microsoft.PowerFx.Connectors/Public/ConnectorRelationships.cs:9]: >public [](http://example.com/codeflow?start=4&length=6)

can we keep internal.  (will make it easier to change later)

File: src/tests/Microsoft.PowerFx.Connectors.Tests.Shared/PublicSurfaceTests.cs [hunk 16]
```diff
@@ -37,9 +37,9 @@ public void PublicSurfaceTest_Connectors()
               "Microsoft.PowerFx.Connectors.ConnectorParameters",
               "Microsoft.PowerFx.Connectors.ConnectorParameterWithSuggestions",
               "Microsoft.PowerFx.Connectors.ConnectorPermission",
+              "Microsoft.PowerFx.Connectors.ConnectorRelationships",
```
Comment (MikeStall) [inline src/tests/Microsoft.PowerFx.Connectors.Tests.Shared/PublicSurfaceTests.cs:40]: >ConnectorRelationships [](http://example.com/codeflow?start=44&length=22)

can we keep these new types internal?

File: src/Directory.Build.props (hunk 9 above)
Comment (LucGenetier) [reply src/Directory.Build.props:34]: I want to be able to run a debug build with warnings
It was like that in the past but recent changes create a "regresssion"
This change isn't related w/ this PR

File: src/tests/Microsoft.PowerFx.Connectors.Tests.Shared/PublicSurfaceTests.cs (hunk 10 above)
Comment (LucGenetier) [reply src/tests/Microsoft.PowerFx.Connectors.Tests.Shared/PublicSurfaceTests.cs:60]: There is no change here (it was public), I just moved these classes in non-Tabular namespace (it wasn't used consistently & not needed) + renamed the class

I can make it internal if you want and we'll see if need it public or not later

File: src/libraries/Microsoft.PowerFx.Connectors/Public/TabularTableType.cs [hunk 17]
```diff
@@ -1,15 +1,16 @@
...
 
+using System;
 using Microsoft.PowerFx.Core.Types;
 using Microsoft.PowerFx.Types;
 
 namespace Microsoft.PowerFx.Connectors
 {
     // Used in ConnectorTableValue
-    internal class ConnectorTableType : TableType
+    public class TabularTableType : TableType
```
[the remark this answers is not in the document]
Comment (LucGenetier) [reply src/libraries/Microsoft.PowerFx.Connectors/Public/TabularTableType.cs:11]: I can make it internal but not remove it due to TabularDType need in TabularIRVisitor - we already had this discussion in a previous PR

File: src/libraries/Microsoft.PowerFx.Connectors/Public/TabularTableValue.cs (hunk 11 above)
Comment (LucGenetier) [reply src/libraries/Microsoft.PowerFx.Connectors/Public/TabularTableValue.cs:16]: I prefer the separation between "Connector" prefix for action connectors and "Tabular" for what's CDP related

File: src/libraries/Microsoft.PowerFx.Connectors/Public/TabularRecordType.cs (hunk 12 above)
Comment (LucGenetier) [reply src/libraries/Microsoft.PowerFx.Connectors/Public/TabularRecordType.cs:55]: I wanted to ensure we log the exception
I'll throw instead

File: src/libraries/Microsoft.PowerFx.Connectors/Public/TabularRecordType.cs (hunk 13 above)
Comment (LucGenetier) [reply src/libraries/Microsoft.PowerFx.Connectors/Public/TabularRecordType.cs:80]: Yes
Equals/GetHashCode are abstract (in AggregateType) and not defined in RecordType
I reused the code from KnownRecordType

File: src/libraries/Microsoft.PowerFx.Connectors/Public/ReferencedEntity.cs (hunk 14 above)
Comment (LucGenetier) [reply src/libraries/Microsoft.PowerFx.Connectors/Public/ReferencedEntity.cs:10]: C# string type

File: src/libraries/Microsoft.PowerFx.Connectors/Public/ConnectorRelationships.cs (hunk 15 above)
Comment (LucGenetier) [reply src/libraries/Microsoft.PowerFx.Connectors/Public/ConnectorRelationships.cs:9]: (I renamed it to TabularRelationships)

This one is intentional as expose relationships in this object which is supposed to be our neutral representation (not SF, ZD, SQL specific) for relationships

File: src/tests/Microsoft.PowerFx.Connectors.Tests.Shared/PublicSurfaceTests.cs (hunk 16 above)
Comment (LucGenetier) [reply src/tests/Microsoft.PowerFx.Connectors.Tests.Shared/PublicSurfaceTests.cs:40]: Not this one for sure, see other comment

File: src/libraries/Microsoft.PowerFx.Connectors/Public/ReferencedEntity.cs (hunk 14 above)
Comment (MikeStall) [reply src/libraries/Microsoft.PowerFx.Connectors/Public/ReferencedEntity.cs:10]: I mean the property (RelationshipName),

File: src/libraries/Microsoft.PowerFx.Connectors/Public/ConnectorRelationships.cs (hunk 15 above)
Comment (MikeStall) [reply src/libraries/Microsoft.PowerFx.Connectors/Public/ConnectorRelationships.cs]: Can we start with just exposing RecordType, and expose this later as we get clearer scenarios?

File: src/libraries/Microsoft.PowerFx.Connectors/Public/TabularRecordType.cs [hunk 18]
```diff
@@ -0,0 +1,74 @@
...
+            if (cr.ExternalTables?.Any() != true)
+            {
+                return base.TryGetFieldType(name, out type);
+            }
+
+            string tableName = cr.ExternalTables.First();
+
+            try
+            {
+                TabularTableDescriptor ttd = TableResolver.ResolveTableAsync(tableName, CancellationToken.None).ConfigureAwait(false).GetAwaiter().GetResult();
```
Comment (MikeStall) [inline src/libraries/Microsoft.PowerFx.Connectors/Public/TabularRecordType.cs:48]: >ResolveTableAsync [](http://example.com/codeflow?start=59&length=17)

ideally - this should be synchronous over a closed set.

File: src/tests/Microsoft.PowerFx.Connectors.Tests.Shared/PowerPlatformTabularTests.cs [hunk 19]
```diff
@@ -471,35 +566,48 @@ public async Task SF_CdpTabular_GetTables()
...
+                "Modified By ID'[User]:s, LastModifiedDate`'Last Modified Date':d, LastReferencedDate`'Last Referenced Date':d, LastViewedDate`'Last Viewed Date':d, MasterRecordId`'Master Record ID'[Account]:s, Name`'Account " +
+                "Name':s, NumberOfEmployees`Employees:w, OwnerId`'Owner ID'[User]:s, ParentId`'Parent Account ID'[Account]:s, Phone`'Account Phone':s, PhotoUrl`'Photo URL':s, ShippingCity`'Shipping City':s, ShippingCountry`'Shipping " +
+                "Country':s, ShippingGeocodeAccuracy`'Shipping Geocode Accuracy':s, ShippingLatitude`'Shipping Latitude':w, ShippingLongitude`'Shipping Longitude':w, ShippingPostalCode`'Shipping Zip/Postal Code':s, ShippingState`'Shipping " +
+                "State/Province':s, ShippingStreet`'Shipping Street':s, SicDesc`'SIC Description':s, SystemModstamp`'System Modstamp':d, Type`'Account Type':s, Website:s]", ((TabularRecordType)sfTable.TabularRecordType).ToStringWithDisplayNames());
+
+            Assert.Equal("Account", sfTable.TabularRecordType.TableSymbolName);
```
Comment (MikeStall) [inline src/tests/Microsoft.PowerFx.Connectors.Tests.Shared/PowerPlatformTabularTests.cs:604]: > Assert.Equal("Account", sfTable.TabularRecordType.TableSymbolName); [](http://example.com/codeflow?start=11&length=68)

:)

**Pull request merged**